### PR TITLE
More flexible PID suffix generation.

### LINKED ISSUE #91: Flexible gGeneration of local names
This looks like a very interesting tool for working with the handle-system. Great idea and excellent docu!

I have a question rearding the unique local name (aka suffix) that is created. From the code it looks like UUID4 strings are always used. Is this correct?

I imagine a use case where one handle-server is used to create PIDs according to different KIPs. The different KIPs define different rules (regular expressions) for the suffix. Validation is probably no problem since the rule is extracted from KIP. But would it be possible to generate conformin PIDs with the "Typed PID Maker"? I can imagine that a plugin would be needed or special code that understands how to generate PIDs given a certain regex-based rule...

### PR DESCRIPTION
**Pull request by Pfeil**
Goal: Make the generation of PID suffixes more flexible. Fixes #91 

Uses decorator pattern to combine suffix properties flexibly.

## Features

- [x] UUIDs (so PIDs may be continued to be created as it is now. Should be default at least until a new major release.)
  - [x] unit tested
- [x] lower/upper case
  - [x] unit tested
- [x] branding
  - [x] unit tested
- [x] fully custom PIDs (trigger is there)
  - [x] unit tested
- [x] Hex-Chunks-Generator with configurable length (amount of chunks)
  - [x] unit tested
- [x] At application start, we derive a PID generator from the given configuration.
- [x] Use the new generator in all systems implemented
    - Note: Consider handling PID generation in front of the PID system/adapter implementations. Use the endpoint level (REST implementation) instead. This should make full customization easier and safer to implement (central point, pid systems only get the record class and get their PID always and only from there).

## Documentation

- [x] All configuration properties are documented in `config/application.properties` and contain sane default values.
- [x] The readme should reflect some possibilities, maybe using examples.
- [x] The webpage should get a PR to have some hint about PID customization. Should it be one of the configuration highlights?
- [x] Developer documentation: Document the structure and some examples of the new PID generator module (e.g. as UML). As the number of features is growing, we need to put more work in this. If this is content for the webpage or should be in the repository somewhere, is yet to be decided.

## Tests

(In addition to usual unit tests and test coverage as enforced by the CI)

- [x] Test the bean generation from the property configuration class.
- [x] Integration test: Custom PIDs shall only work if allowed in the configuration.

### CODE REVIEW

File: src/main/java/edu/kit/datamanager/pit/pidgeneration/generators/PidSuffixGenUpperCase.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package edu.kit.datamanager.pit.pidgeneration.generators;
+
+import edu.kit.datamanager.pit.pidgeneration.PidSuffix;
+import edu.kit.datamanager.pit.pidgeneration.PidSuffixGenerator;
+
+/**
+ * Generates a PID suffix based on a contained generator and returns the result
+ * in lower case.
```
Comment (dalito) [inline src/main/java/edu/kit/datamanager/pit/pidgeneration/generators/PidSuffixGenUpperCase.java:8]: should be "upper case" (also in `PidSuffixGenPrefixed.java`)

Comment (Pfeil) [reply src/main/java/edu/kit/datamanager/pit/pidgeneration/generators/PidSuffixGenUpperCase.java]: Thank you :) fixed.

**Comment by Pfeil** [comment]
For reference when merging master into #114 later: I took some code from #114 for the tests. Merging should be easy:

- The MockApiUtils can be fully used from #114
- The CustomPidTest has been copied and some tests been added. The original test has been modified only slightly.

---

General status: Everything seems to be in place now. The rest is mostly documentation and the actual implementation of something more recognizable, distinguishable, less scary, short(?), and/or nameable.

File: src/main/java/edu/kit/datamanager/pit/pidgeneration/generators/HexChunksGenerator.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import edu.kit.datamanager.pit.pidgeneration.PidSuffixGenerator;
+
+/**
+ * Generates hex chunks of four characters each, separated by a dash.
+ * 
+ * Corresponds to four chunks by default, offering more than enough PIDs.
+ */
+public class HexChunksGenerator implements PidSuffixGenerator {
+
+    protected Random random = new Random();
```
[the remark this answers is not in the document]
Comment (Pfeil) [reply src/main/java/edu/kit/datamanager/pit/pidgeneration/generators/HexChunksGenerator.java:17]: We are not in a security critical context here. As long as it is not obviously deterministic, it is fine. I'll double-check this and consider using a better generator later.

**Pull request merged**